search/includes/classes/class-concurrency-limiter.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Automattic\VIP\Search;
+
+use WP_Error;
+
+class Concurrency_Limiter {
+	const KEY_NAME = 'vip_es_request_count';
+
+	/** @var int */
+	private $cache_ttl = 60;
+
+	/** @var int */
+	private $max_concurrent_requests = 100;
```

Comment:
These values are to be discussed.

Reply:
This seems kinda low... I wonder how to come up with optimal value.. Do we know how many concurent request clients do currently?

Reply:
Unfortunately, I have no idea. I am open to changes here :-)

And, for corner cases, we provide a filter to adjust this value.

Reply:
PS: this is a per-pod limit.

Reply:
Since it's per-pod, and PHP is basically always synchronous, I'm not sure how much benefit there will be here since a pod already has a concurrency limit built-in with the "php workers" limit. This is a pretty standard number across the board, except in rare cases of manual vertical scaling rather than horizontal.

Reply:
>  Do we know how many concurent request clients do currently?

I don't think we have a good understanding of this currently. I think the first step would be to gauge that value by disabling the limiting functionality and instead just log the fact. We can keep it running for a bit to gather a better understanding.

Reply:
I have created another backend (in #2706) that will use Memcached, and then this limit will be per-client.

For the per-pod limit, I would like to confirm if we are not using php-fpm's dynamic process scheduler (it spawns new PHP workers on demand and kills them when they are not in use).

Reply:
I have updated #2706 to use Object Cache (= Memcached). This makes the limit global (per-client instead of per-pod).

Also, that PR changes the behavior to log warnings for excessive requests instead of dropping them.